[aotd] coerce_same_metadata_as_tangent with expected_type for e.g.AsyncCollectiveTensor

IvanKobzarev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #139095

Based on discussion here: https://github.com/pytorch/pytorch/pull/138731

Introducing ability for subclass implement type convertion to expected_type.
```
    def __coerce_same_metadata_as_tangent__(
        self, expected_metadata: Any, expected_type: Optional[Type] = None
    ):
```
Here if `expected_type=None` means `SubclassClass` is expected.

E.g. for `DTensor` we may find tangent `AsyncCollectiveTensor` where we expected `Tensor` - in this case 
`expected_type=Tensor` will be called during runtime

Adding implementation to AsyncCollectiveTensor, that just triggers `wait()`.


cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

IvanKobzarev committed:
[aotd] coerce_same_metadata_as_tangent with expected_type for e.g.AsyncCollectiveTensor

[ghstack-poisoned]

bdhirsh commented on torch/_functorch/_aot_autograd/runtime_wrappers.py:
```diff
@@ -1446,6 +1446,12 @@ def process_runtime_tangent(x, meta: Union[PlainTensorMeta, SubclassCreationMeta
+        if is_subclass and not is_subclass_meta:
+            # Unexpected subclass, during tracing we guessed it was a plain Tensor
+            if hasattr(x, "__coerce_same_metadata_as_tangent__"):
+                x = x.__coerce_same_metadata_as_tangent__(None, torch.Tensor)
```
hmm. Two things:

(1) Right now this branch is hardcoded for the case where we expected a subclass tangent but got a plain tensor tangent. But we specifically updated the coerce API to be more generic: it can technically allow a subclass to convert to any other subclass type, if they are able to handle it. It seems to me like if we are going with that more general API, we should properly handle that here: don't haredcode `torch.Tensor`, just directly pass in the `type(x)`.

Since we're potentially being BC breaking (I think only DTensor uses this API, although I think a few people have forked DTensor out of tree over time), it might also be better to optionally only pass the type argument in when the types are different

(2) you added this call as a completely new one, on top of the existing a call below (`x = x.__coerce_same_metadata_as_tangent__(meta.meta)`). It seems better to consolidate them into a single call?

IvanKobzarev committed:
Update on "[aotd] coerce_same_metadata_as_tangent with expected_type for e.g.AsyncCollectiveTensor"


Based on discussion here: https://github.com/pytorch/pytorch/pull/138731

Introducing ability for subclass implement type convertion to expected_type.
```
    def __coerce_same_metadata_as_tangent__(
        self, expected_metadata: Any, expected_type: Optional[Type] = None
    ):
```
Here if `expected_type=None` means `SubclassClass` is expected.

E.g. for `DTensor` we may find tangent `AsyncCollectiveTensor` where we expected `Tensor` - in this case 
`expected_type=Tensor` will be called during runtime

Adding implementation to AsyncCollectiveTensor, that just triggers `wait()`.


cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] coerce_same_metadata_as_tangent with expected_type for e.g.AsyncCollectiveTensor"


Based on discussion here: https://github.com/pytorch/pytorch/pull/138731

Introducing ability for subclass implement type convertion to expected_type.
```
    def __coerce_same_metadata_as_tangent__(
        self, expected_metadata: Any, expected_type: Optional[Type] = None
    ):
```
Here if `expected_type=None` means `SubclassClass` is expected.

E.g. for `DTensor` we may find tangent `AsyncCollectiveTensor` where we expected `Tensor` - in this case 
`expected_type=Tensor` will be called during runtime

Adding implementation to AsyncCollectiveTensor, that just triggers `wait()`.


cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

bdhirsh commented on torch/distributed/tensor/_api.py:
```diff
@@ -325,7 +325,8 @@ def __coerce_tangent_metadata__(self):
 
-    def __coerce_same_metadata_as_tangent__(self, flatten_spec):
+    def __coerce_same_metadata_as_tangent__(self, flatten_spec, expected_type=None):
+        assert expected_type is None
```
Nit: this can be user facing, if the user ends up running their code in such a way that e.g. the expected tangents are DTensors but the actual tangents are plain tensors. So we should make sure the error message if this assert fails is very clear. (include the actual/expected types and metadata)

bdhirsh replied:
Or alternatively, we could have this function return `None` if it is unable to properly coerce, and let AOTAutograd  raise the error itself if it sees a `None` return

IvanKobzarev replied:
Yes, with current logic if it returns None we will raise an error with all details. https://github.com/pytorch/pytorch/blob/main/torch/_functorch/_aot_autograd/runtime_wrappers.py#L1486
I think we should also document the logic of coercing by metadata and type, the meaning of None etc.

bdhirsh replied:
hmm this code as-written (the assert above) still seem like it will regress our error message?

Prior to this PR, if we expected a `DTensor` tangent but we got a plain tensor tangent, we would get this nice error message: https://github.com/pytorch/pytorch/blob/main/torch/_functorch/_aot_autograd/runtime_wrappers.py#L1505

Seems like you want to return `None` here?

IvanKobzarev replied:
Yes, better to not assert and just return `None` here for user-friendly error message.

bdhirsh commented on torch/distributed/_functional_collectives.py:
```diff
@@ -600,6 +600,17 @@ def __tensor_unflatten__(inner_tensors, meta, outer_size, outer_stride):
+            raise RuntimeError("Not implemented")
+
+        t = self.trigger_wait()
+        while isinstance(t, AsyncCollectiveTensor):
```
nit: we shouldn't need this while loop either (I think we will be a in a very weird place if there are every nested AsyncCollectiveTensors, so it seems pointless to defensively program for it)

bdhirsh commented on torch/_functorch/_aot_autograd/runtime_wrappers.py:
```diff
@@ -1440,48 +1440,61 @@ class AutogradLazyBackwardCompileInfo:
-
-        if not x.is_contiguous(memory_format=meta.memory_format):
-            x = x.contiguous(memory_format=meta.memory_format)
+        if is_subclass and not is_subclass_meta:
```
Ok, concretely, I think:

(1) instead of special-casing the is_subclass case, we should just check if the type of the two tensors is different, and if so then unconditionally call `coerce(metadata or None, type(actual))`

(2) if the `coerce()` function returns None, the subclass has indicated that it cannot perform the coercion, and so we can raise the old error message (with even more useful information in the error than the subclass could have given, like exactly which tangent in the tangent list we are up to)

IvanKobzarev replied:
Ok. Fused the logic in one coercion and updated PR.

IvanKobzarev commented on torch/_functorch/_aot_autograd/runtime_wrappers.py:
```diff
@@ -1440,48 +1440,61 @@ class AutogradLazyBackwardCompileInfo:
+        def maybe_coerce_to_memory_format(t, memory_format):
+            if not t.is_contiguous(memory_format=meta.memory_format):
+                return t.contiguous(memory_format=meta.memory_format)
+            return t
```
maybe_coerce_to_memory_format  is only for memory_format processing of one argument.

The whole function `process_runtime_tangent` returns tuple to be able to do flatenning (2d item in tuple) at the same traversal as processing, to do only one traversal in total.
So the return type is `Tuple[ChangedRawItem, FlattenedChangedItems]`

IvanKobzarev committed:
Update on "[aotd] coerce_same_metadata_as_tangent with expected_type for e.g.AsyncCollectiveTensor"


Based on discussion here: https://github.com/pytorch/pytorch/pull/138731

Introducing ability for subclass implement type convertion to expected_type.
```
    def __coerce_same_metadata_as_tangent__(
        self, expected_metadata: Any, expected_type: Optional[Type] = None
    ):
```
Here if `expected_type=None` means `SubclassClass` is expected.

E.g. for `DTensor` we may find tangent `AsyncCollectiveTensor` where we expected `Tensor` - in this case 
`expected_type=Tensor` will be called during runtime

Adding implementation to AsyncCollectiveTensor, that just triggers `wait()`.


cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] coerce_same_metadata_as_tangent with expected_type for e.g.AsyncCollectiveTensor"


Based on discussion here: https://github.com/pytorch/pytorch/pull/138731

Introducing ability for subclass implement type convertion to expected_type.
```
    def __coerce_same_metadata_as_tangent__(
        self, expected_metadata: Any, expected_type: Optional[Type] = None
    ):
```
Here if `expected_type=None` means `SubclassClass` is expected.

E.g. for `DTensor` we may find tangent `AsyncCollectiveTensor` where we expected `Tensor` - in this case 
`expected_type=Tensor` will be called during runtime

Adding implementation to AsyncCollectiveTensor, that just triggers `wait()`.


cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] coerce_same_metadata_as_tangent with expected_type for e.g.AsyncCollectiveTensor"


Based on discussion here: https://github.com/pytorch/pytorch/pull/138731

Introducing ability for subclass implement type convertion to expected_type.
```
    def __coerce_same_metadata_as_tangent__(
        self, expected_metadata: Any, expected_type: Optional[Type] = None
    ):
```
Here if `expected_type=None` means `SubclassClass` is expected.

E.g. for `DTensor` we may find tangent `AsyncCollectiveTensor` where we expected `Tensor` - in this case 
`expected_type=Tensor` will be called during runtime

Adding implementation to AsyncCollectiveTensor, that just triggers `wait()`.


cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] coerce_same_metadata_as_tangent with expected_type for e.g.AsyncCollectiveTensor"


Based on discussion here: https://github.com/pytorch/pytorch/pull/138731

Introducing ability for subclass implement type convertion to expected_type.
```
    def __coerce_same_metadata_as_tangent__(
        self, expected_metadata: Any, expected_type: Optional[Type] = None
    ):
```
Here if `expected_type=None` means `SubclassClass` is expected.

E.g. for `DTensor` we may find tangent `AsyncCollectiveTensor` where we expected `Tensor` - in this case 
`expected_type=Tensor` will be called during runtime

Adding implementation to AsyncCollectiveTensor, that just triggers `wait()`.


cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] coerce_same_metadata_as_tangent with expected_type for e.g.AsyncCollectiveTensor"


Based on discussion here: https://github.com/pytorch/pytorch/pull/138731

Introducing ability for subclass implement type convertion to expected_type.
```
    def __coerce_same_metadata_as_tangent__(
        self, expected_metadata: Any, expected_type: Optional[Type] = None
    ):
```
Here if `expected_type=None` means `SubclassClass` is expected.

E.g. for `DTensor` we may find tangent `AsyncCollectiveTensor` where we expected `Tensor` - in this case 
`expected_type=Tensor` will be called during runtime

Adding implementation to AsyncCollectiveTensor, that just triggers `wait()`.


cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] coerce_same_metadata_as_tangent with expected_type for e.g.AsyncCollectiveTensor"


Based on discussion here: https://github.com/pytorch/pytorch/pull/138731

Introducing ability for subclass implement type convertion to expected_type.
```
    def __coerce_same_metadata_as_tangent__(
        self, expected_metadata: Any, expected_type: Optional[Type] = None
    ):
```
Here if `expected_type=None` means `SubclassClass` is expected.

E.g. for `DTensor` we may find tangent `AsyncCollectiveTensor` where we expected `Tensor` - in this case 
`expected_type=Tensor` will be called during runtime

Adding implementation to AsyncCollectiveTensor, that just triggers `wait()`.


cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] coerce_same_metadata_as_tangent with expected_type for e.g.AsyncCollectiveTensor"


Based on discussion here: https://github.com/pytorch/pytorch/pull/138731

Introducing ability for subclass implement type convertion to expected_type.
```
    def __coerce_same_metadata_as_tangent__(
        self, expected_metadata: Any, expected_type: Optional[Type] = None
    ):
```
Here if `expected_type=None` means `SubclassClass` is expected.

E.g. for `DTensor` we may find tangent `AsyncCollectiveTensor` where we expected `Tensor` - in this case 
`expected_type=Tensor` will be called during runtime

Adding implementation to AsyncCollectiveTensor, that just triggers `wait()`.


cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] coerce_same_metadata_as_tangent with expected_type for e.g.AsyncCollectiveTensor"


Based on discussion here: https://github.com/pytorch/pytorch/pull/138731

Introducing ability for subclass implement type convertion to expected_type.
```
    def __coerce_same_metadata_as_tangent__(
        self, expected_metadata: Any, expected_type: Optional[Type] = None
    ):
```
Here if `expected_type=None` means `SubclassClass` is expected.

E.g. for `DTensor` we may find tangent `AsyncCollectiveTensor` where we expected `Tensor` - in this case 
`expected_type=Tensor` will be called during runtime

Adding implementation to AsyncCollectiveTensor, that just triggers `wait()`.


cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] coerce_same_metadata_as_tangent with expected_type for e.g.AsyncCollectiveTensor"


Based on discussion here: https://github.com/pytorch/pytorch/pull/138731

Introducing ability for subclass implement type convertion to expected_type.
```
    def __coerce_same_metadata_as_tangent__(
        self, expected_metadata: Any, expected_type: Optional[Type] = None
    ):
```
Here if `expected_type=None` means `SubclassClass` is expected.

E.g. for `DTensor` we may find tangent `AsyncCollectiveTensor` where we expected `Tensor` - in this case 
`expected_type=Tensor` will be called during runtime

Adding implementation to AsyncCollectiveTensor, that just triggers `wait()`.


cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] coerce_same_metadata_as_tangent with expected_type for e.g.AsyncCollectiveTensor"


Based on discussion here: https://github.com/pytorch/pytorch/pull/138731

Introducing ability for subclass implement type convertion to expected_type.
```
    def __coerce_same_metadata_as_tangent__(
        self, expected_metadata: Any, expected_type: Optional[Type] = None
    ):
```
Here if `expected_type=None` means `SubclassClass` is expected.

E.g. for `DTensor` we may find tangent `AsyncCollectiveTensor` where we expected `Tensor` - in this case 
`expected_type=Tensor` will be called during runtime

Adding implementation to AsyncCollectiveTensor, that just triggers `wait()`.


cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] coerce_same_metadata_as_tangent with expected_type for e.g.AsyncCollectiveTensor"


Based on discussion here: https://github.com/pytorch/pytorch/pull/138731

Introducing ability for subclass implement type convertion to expected_type.
```
    def __coerce_same_metadata_as_tangent__(
        self, expected_metadata: Any, expected_type: Optional[Type] = None
    ):
```
Here if `expected_type=None` means `SubclassClass` is expected.

E.g. for `DTensor` we may find tangent `AsyncCollectiveTensor` where we expected `Tensor` - in this case 
`expected_type=Tensor` will be called during runtime

Adding implementation to AsyncCollectiveTensor, that just triggers `wait()`.


cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

bdhirsh commented on torch/testing/_internal/subclasses.py:
```diff
@@ -0,0 +1,76 @@
+from torch.utils._python_dispatch import return_and_correct_aliasing
+
+
+class WrapSC(torch.Tensor):
```
nit: up to you, but I find `WrapperSubclass` a bit clearer than `WrapperSC`

bdhirsh commented on torch/testing/_internal/subclasses.py:
```diff
@@ -0,0 +1,76 @@
+    def __tensor_unflatten__(inner_tensors, meta, outer_size, outer_stride):
+        assert meta is None
+        a = inner_tensors["a"]
+        if type(a) is torch.Tensor:
```
I guess the purpose of this check is that you only want to run the assertions below if we are at runtime, not trace time.

This isn't very robust though, since it will fail if the inner tensor `a` is itself another subclass.

You probably want to use `is_fake(a)`, to tell if you are in the middle of tracing?

IvanKobzarev replied:
Honestly, I just copied this from TwoTensor :)

bdhirsh commented on torch/testing/_internal/subclasses.py:
```diff
@@ -0,0 +1,76 @@
+        self, expected_metadata: Any, expected_type: Optional[Type] = None
+    ):
+        if expected_type is torch.Tensor:
+            return self.a
```
tbh this code as written is a bit confusing: the idea of this function is that it is supposed to **enforce** that the return type is the same as `expected_type`. You might want to add an `assert type(self.a) == expected_type` here?

IvanKobzarev replied:
Thanks, yes, this will be more general to be able to wrap subclasses and coerce to them.

IvanKobzarev committed:
Update on "[aotd] coerce_same_metadata_as_tangent with expected_type for e.g.AsyncCollectiveTensor"


Based on discussion here: https://github.com/pytorch/pytorch/pull/138731

Introducing ability for subclass implement type convertion to expected_type.
```
    def __coerce_same_metadata_as_tangent__(
        self, expected_metadata: Any, expected_type: Optional[Type] = None
    ):
```
Here if `expected_type=None` means `SubclassClass` is expected.

E.g. for `DTensor` we may find tangent `AsyncCollectiveTensor` where we expected `Tensor` - in this case 
`expected_type=Tensor` will be called during runtime

Adding implementation to AsyncCollectiveTensor, that just triggers `wait()`.


cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

bdhirsh reviewed (approved):
lgtm

IvanKobzarev commented:
@pytorchbot merge
